packages/core/src/browser/shell/view-contribution.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable, inject } from "inversify";
+import {
+    MenuModelRegistry, Command, CommandContribution,
+    MenuContribution, KeybindingContribution, KeybindingRegistry,
+    KeyCode, CommandRegistry
+} from '../../common';
+import { WidgetManager } from '../widget-manager';
+import { Widget } from '@phosphor/widgets';
+import { FrontendApplicationContribution, FrontendApplication } from '../frontend-application';
+import { CommonMenus } from '../common-frontend-contribution';
+import { ApplicationShell } from './application-shell';
+
+export interface OpenViewArguments extends ApplicationShell.WidgetOptions {
+    toggle?: boolean
+    activate?: boolean;
+}
+
+export interface ViewContributionOptions {
+    widgetId: string;
+    widgetName: string;
+    defaultWidgetOptions: ApplicationShell.WidgetOptions;
+    toggleCommandId?: string;
+    toggleKeybinding?: KeyCode;
+}
+
+/**
+ * An abstract superclass for frontend contributions that add a view to the application shell.
+ */
+@injectable()
+export abstract class AbstractViewContribution<T extends Widget> implements CommandContribution, MenuContribution, KeybindingContribution, FrontendApplicationContribution {
+
+    @inject(WidgetManager) protected widgetManager: WidgetManager;
+    @inject(ApplicationShell) protected shell: ApplicationShell;
+
+    readonly toggleCommand?: Command;
+
+    constructor(
+        protected readonly options: ViewContributionOptions
+    ) {
+        if (options.toggleCommandId) {
+            this.toggleCommand = {
+                id: options.toggleCommandId,
+                label: 'Toggle ' + options.widgetName + ' View'
+            };
+        }
+    }
+
+    get widget(): Promise<T> {
+        return this.widgetManager.getOrCreateWidget<T>(this.options.widgetId);
+    }
+
+    async openView(args: Partial<OpenViewArguments> = {}): Promise<T> {
+        const widget = await this.widget;
+        const tabBar = this.shell.getTabBarFor(widget);
+        if (!tabBar) {
+            // The widget is not attached yet, so add it to the shell
+            const widgetArgs: OpenViewArguments = {
+                ...this.options.defaultWidgetOptions,
+                ...args
+            };
+            this.shell.addWidget(widget, widgetArgs);
+        } else if (tabBar.currentTitle === widget.title && args.toggle) {
```

Comment:
`widget` has `isVisible` attribute, could it be used here?

Reply:
or `this.shell.currentWidget === widget` it should mean that a widget is attached, visible and in focus

Reply:
The widget should be closed when it is already visible in order to realize the toggle behavior. The tab bars indicate which widgets are visible.